README.md
This module implements a python `logging` formatter which produces well-formed RFC5424-compatible Syslog messages to a given socket.

[![CI](https://github.com/EasyPost/syslog-rfc5424-formatter/workflows/CI/badge.svg)](https://github.com/EasyPost/syslog-rfc5424-formatter/actions?query=workflow%3ACI)
[![PyPI version](https://badge.fury.io/py/syslog-rfc5424-formatter.svg)](https://badge.fury.io/py/syslog-rfc5424-formatter)
[![Documentation Status](https://readthedocs.org/projects/syslog-rfc5424-formatter/badge/?version=latest)](https://syslog-rfc5424-formatter.readthedocs.io/en/latest/?badge=latest)


## Usage

If you're configuring your loggers from code, you should use this formatter as below:

```python
import logging
import logging.handlers
from syslog_rfc5424_formatter import RFC5424Formatter


def set_up_logging():
    h = logging.handlers.SysLogHandler('/path/to/syslog_socket')
    h.setFormatter(RFC5424Formatter())
    logging.getLogger('').addHandler(h)
```


If you're using a more modern combination of a JSON/YAML config file and `logging.config.dictConfig`, your config file should look like the following (assuming YAML concrete syntax):

```yaml
formatters:
    syslog:
        (): syslog_rfc5424_formatter.RFC5424Formatter

handlers:
    syslog:
        formatter: syslog
        class: logging.handlers.SysLogHandler
        address: "/path/to/syslog/socket"
        facility: "ext://logging.handlers.SysLogHandler.LOG_USER"

root:
    level: INFO
    handlers:
        - syslog
```

## License

This work is licensed under the ISC license, the text of which can be found at [LICENSE.txt](LICENSE.txt).


CHANGES.md
1.2.3
----
- Drop support for Python 3.3, 3.4, and 3.5
- Move CI from Travis to Github Actions
- Improve packaging

1.2.2
-----
- Fix a bug when using records with arguments

1.2.1
-----
- Raise an exception if you pass structured data without setting SD-ID

1.2.0
-----
- Support overriding procid/msgid and setting structured data (#11; thanks @tkoopman)
- Add some new tests

1.1.1
-----
- Add some more meta-information to setup.py
- Set up ReadTheDocs

1.1.0
-----
- Drop support for Python 3.3
- Add support for Python 3.7

1.0.1
-----
Clean up some stuff in `setup.py`; no user-visible changes


tests/test_syslog_rfc5424_formatter.py
import logging
import logging.handlers
import tempfile
import shutil
import os
import socket
import time

import pytest
from unittest import TestCase

import mock
from syslog_rfc5424_parser import SyslogMessage
from syslog_rfc5424_parser.constants import SyslogSeverity, SyslogFacility

from syslog_rfc5424_formatter import RFC5424Formatter, InvalidSDIDError


@mock.patch('socket.gethostname', return_value='the_host')
@mock.patch('os.getpid', return_value=1)
@mock.patch('time.time', return_value=0)
@mock.patch.dict(os.environ, {'TZ': 'UTC'})
class RFC5424FormatterTestCase(TestCase):
    def setUp(self):
        time.tzset()

    def test_basic(*args):
        f = RFC5424Formatter()
        r = logging.makeLogRecord({'name': 'root', 'msg': 'A Message'})
        assert f.format(r) == '1 1970-01-01T00:00:00Z the_host root 1 - - A Message'

    def test_non_utc(*args):
        with mock.patch.dict(os.environ, {'TZ': 'America/Phoenix'}):
            time.tzset()
            f = RFC5424Formatter()
            r = logging.makeLogRecord({'name': 'root', 'msg': 'A Message'})
            assert f.format(r) == '1 1969-12-31T17:00:00-07:00 the_host root 1 - - A Message'

    def test_long_pid(*args):
        with mock.patch('os.getpid', return_value=999999):
            f = RFC5424Formatter()
            r = logging.makeLogRecord({'name': 'root', 'msg': 'A Message'})
            assert f.format(r) == '1 1970-01-01T00:00:00Z the_host root 999999 - - A Message'

    def test_properties(*args):
        f = RFC5424Formatter()
        f.msgid = 'msgid'
        f.procid = 999999
        r = logging.makeLogRecord({'name': 'root', 'msg': 'A Message'})
        assert f.format(r) == '1 1970-01-01T00:00:00Z the_host root 999999 msgid - A Message'

    def test_properties_override(*args):
        f = RFC5424Formatter()
        f.msgid = 'msgid'
        f.procid = 999999
        r = logging.makeLogRecord({'name': 'root', 'msg': 'A Message', 'args': {'msgid': 'digsm', 'procid': 1234}})
        assert f.format(r) == '1 1970-01-01T00:00:00Z the_host root 1234 digsm - A Message'

    def test_structured_data(*args):
        f = RFC5424Formatter(sd_id='Undefined@32473')
        r = logging.makeLogRecord({'name': 'root', 'msg': 'A Message', 'args': {
            'structured_data': {'name': 'value', 'sdid@32473': {'escape': '\\"]'}}
        }})
        result = f.format(r)
        assert '[sdid@32473 escape="\\\\\\"\\]"]' in result
        assert '[Undefined@32473 name="value"]' in result
        f = RFC5424Formatter()
        f.sd_id = 'Undefined@32474'
        r = logging.makeLogRecord({'name': 'root', 'msg': 'A Message', 'args': {
            'structured_data': {'name': 'value', 'sdid@32474': {'escape': '\\"]'}}
        }})
        result = f.format(r)
        assert '[sdid@32474 escape="\\\\\\"\\]"]' in result
        assert '[Undefined@32474 name="value"]' in result

    def test_structured_data_no_sdid(*args):
        f = RFC5424Formatter()
        r = logging.makeLogRecord({'name': 'root', 'msg': 'A Message', 'args': {
            'structured_data': {'name': 'value', 'sdid@32473': {'escape': '\\"]'}}
        }})
        with pytest.raises(InvalidSDIDError):
            f.format(r)

    def test_format_string(*args):
        f = RFC5424Formatter('%(message)s banana')
        r = logging.makeLogRecord({'name': 'root', 'msg': 'A Message'})
        assert f.format(r) == '1 1970-01-01T00:00:00Z the_host root 1 - - A Message banana'

    def test_integration(*args):
        try:
            working_dir = tempfile.mkdtemp()
            s_path = os.path.join(working_dir, 'sock')
            s = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
            s.bind(s_path)
            r = logging.makeLogRecord({'name': 'root', 'msg': 'A Message', 'levelname': 'DEBUG'})
            h = logging.handlers.SysLogHandler(s_path, facility=logging.handlers.SysLogHandler.LOG_USER)
            h.setFormatter(RFC5424Formatter())
            h.handle(r)
            s.settimeout(1)
            msg_body = s.recv(1024)
            assert msg_body == b'<15>1 1970-01-01T00:00:00Z the_host root 1 - - A Message\x00'
            fields = SyslogMessage.parse(msg_body.decode('utf-8'))
            assert fields.severity == SyslogSeverity.debug
            assert fields.facility == SyslogFacility.user
            assert fields.version == 1
            assert fields.hostname == 'the_host'
            assert fields.appname == 'root'
            assert fields.msg == 'A Message\x00'
            assert fields.procid == 1
        finally:
            s.close()
            shutil.rmtree(working_dir)

    def test_integration_with_structured_data(*args):
        try:
            working_dir = tempfile.mkdtemp()
            s_path = os.path.join(working_dir, 'sock')
            s = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
            s.bind(s_path)
            r = logging.makeLogRecord({'name': 'root', 'msg': 'A Message', 'levelname': 'DEBUG', 'args': {
                'structured_data': {'foo': '1', 'baz': 2},
            }})
            h = logging.handlers.SysLogHandler(s_path, facility=logging.handlers.SysLogHandler.LOG_USER)
            h.setFormatter(RFC5424Formatter(sd_id='foobar'))
            h.handle(r)
            s.settimeout(1)
            msg_body = s.recv(1024)
            fields = SyslogMessage.parse(msg_body.decode('utf-8'))
            assert fields.msg == 'A Message\x00'
            assert fields.sd == {'foobar': {'foo': '1', 'baz': '2'}}
        finally:
            s.close()
            shutil.rmtree(working_dir)

    def test_hostname_fails(*args):
        with mock.patch('socket.gethostname', side_effect=ValueError):
            f = RFC5424Formatter()
            r = logging.makeLogRecord({'name': 'root', 'msg': 'A Message'})
            assert f.format(r) == '1 1970-01-01T00:00:00Z - root 1 - - A Message'

    def test_logger_integration(*args):
        try:
            working_dir = tempfile.mkdtemp()
            s_path = os.path.join(working_dir, 'sock')
            s = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
            s.bind(s_path)
            h = logging.handlers.SysLogHandler(s_path, facility=logging.handlers.SysLogHandler.LOG_USER)
            h.setFormatter(RFC5424Formatter())
            lr = logging.getLogger('test_logger_integration')
            lr.addHandler(h)
            lr.setLevel(logging.DEBUG)
            lr.info('this is a test %s %d', 'foo', 1)
            s.settimeout(1)
            msg_body = s.recv(1024)
            assert msg_body == b'<14>1 1970-01-01T00:00:00Z the_host test_logger_integration 1 - - this is a test foo 1\x00'  # noqa
        finally:
            s.close()
            shutil.rmtree(working_dir)


syslog_rfc5424_formatter.py
import logging
import time
import socket
import datetime
import re

version_info = (1, 2, 2)
__version__ = '.'.join(str(s) for s in version_info)
__author__ = 'EasyPost <oss@easypost.com>'


class RFC5424FormatterError(Exception):
    pass


class InvalidSDIDError(RFC5424FormatterError):
    pass


class RFC5424Formatter(logging.Formatter, object):
    '''
    A derived formatter than allows for isotime specification
    for full RFC5424 compliancy (with corrected TZ format).

    This should be combined with the Syslog Handler to actually emit logs.

    For a "proper" ISOTIME format, use "%(isotime)s" in a
    formatter instance of this class or a class derived from
    this class.  This is for a work-around where strftime
    has no mechanism to produce timezone in the format of
    "-08:00" as required by RFC5424.

    The '%(isotime)s' replacement will read in the record
    timestamp and try and reparse it.  This really is a
    problem with RFC5424 and strftime.  I am unsure if this
    will be fixed in the future (in one or the other case)

    This formatter has an added benefit of allowing for
    '%(hostname)s' to be specified which will return a '-'
    as specified in RFC5424 if socket.gethostname() returns
    bad data (exception).

    This formatter will automatically insert the RFC5424 header for you;
    the format string that you pass in the constructor is only
    applied to the message body (and should typically just be %(message)).

    Stuctured Data Example:

        [exampleSDID@32473 iut="3" eventSource="Application" eventID="1011"]

    To use structured data:

       1. Construct the logger with an sd_id kwarg (or set the `sd_id` attribute on the logger object)
       2. Construct your individual records with `{'args': {'structured_data': {'iut': '3'}}}`
    '''
    def __init__(self, fmt='%(message)s', datefmt=None, style='%', procid=None, msgid=None, sd_id=None):
        # note: we accept and throw away "style" because our stuff only works with %
        # we also accept and throw away datefmt for similar reasons
        self._tz_fix = re.compile(r'([+-]\d{2})(\d{2})$')
        self._procid = procid
        self._msgid = msgid
        self._sd_id = sd_id
        return super(RFC5424Formatter, self).__init__(fmt=fmt, datefmt=None)

    @property
    def procid(self):
        """Default PROCID to add to syslog message"""
        return self._procid

    @procid.setter
    def procid(self, id):
        self._procid = id

    @property
    def msgid(self):
        """Default MSGID to add to syslog message"""
        return self._msgid

    @msgid.setter
    def msgid(self, id):
        self._msgid = id

    @property
    def sd_id(self):
        """Default SD-ID to add to STRUCTURED-DATA section in syslog message"""
        return self._sd_id

    @sd_id.setter
    def sd_id(self, sd_id):
        if not sd_id:
            raise InvalidSDIDError('SD-ID cannot be empty')
        self._sd_id = sd_id

    def format(self, record):
        try:
            record.__dict__['hostname'] = socket.gethostname()
        except Exception:
            record.__dict__['hostname'] = '-'
        isotime = datetime.datetime.fromtimestamp(record.created).isoformat()
        tz = self._tz_fix.match(time.strftime('%z'))
        if time.timezone and tz:
            (offset_hrs, offset_min) = tz.groups()
            if int(offset_hrs) == 0 and int(offset_min) == 0:
                isotime = isotime + 'Z'
            else:
                isotime = '{0}{1}:{2}'.format(isotime, offset_hrs, offset_min)
        else:
            isotime = isotime + 'Z'

        record.__dict__['isotime'] = isotime
        record.__dict__['procid'] = self.procid if self.procid else record.process
        record.__dict__['msgid'] = self.msgid if self.msgid else '-'

        if 'structured_data' in record.args:
            if not isinstance(record.args['structured_data'], dict):
                raise ValueError('structured_data must be a dict')

            all_sddata = {}
            default_sdparam = {}

            for key, value in record.args['structured_data'].items():
                if isinstance(value, dict):
                    all_sddata[key] = value
                else:
                    default_sdparam[key] = value

            if len(default_sdparam) > 0:
                if self.sd_id in all_sddata:
                    raise InvalidSDIDError('Cannot use same SD-ID twice')
                if not self.sd_id:
                    raise InvalidSDIDError('SD-ID cannot be empty')
                all_sddata[self.sd_id] = default_sdparam

            sd = ''
            for sdid, data in all_sddata.items():
                sd += '[{0}'.format(sdid)
                for key, value in data.items():
                    escaped = str(value).replace('\\', '\\\\').replace(']', '\\]').\
                        replace('"', '\\"')
                    sd += ' {0}="{1}"'.format(key, escaped)
                sd += ']'

            record.__dict__['sd'] = sd
        else:
            record.__dict__['sd'] = '-'

        for key in ('procid', 'msgid'):
            if key in record.args:
                record.__dict__[key] = record.args.pop(key)

        header = '1 {isotime} {hostname} {name} {procid} {msgid} {sd} '.format(
            **record.__dict__
        )
        return header + super(RFC5424Formatter, self).format(record)


syslog_rfc5424_formatter/__init__.py
import logging
import time
import socket
import datetime
import re

version_info = (1, 2, 3)
__version__ = '.'.join(str(s) for s in version_info)
__author__ = 'EasyPost <oss@easypost.com>'


class RFC5424FormatterError(Exception):
    pass


class InvalidSDIDError(RFC5424FormatterError):
    pass


class RFC5424Formatter(logging.Formatter, object):
    '''
    A derived formatter than allows for isotime specification
    for full RFC5424 compliancy (with corrected TZ format).

    This should be combined with the Syslog Handler to actually emit logs.

    For a "proper" ISOTIME format, use "%(isotime)s" in a
    formatter instance of this class or a class derived from
    this class.  This is for a work-around where strftime
    has no mechanism to produce timezone in the format of
    "-08:00" as required by RFC5424.

    The '%(isotime)s' replacement will read in the record
    timestamp and try and reparse it.  This really is a
    problem with RFC5424 and strftime.  I am unsure if this
    will be fixed in the future (in one or the other case)

    This formatter has an added benefit of allowing for
    '%(hostname)s' to be specified which will return a '-'
    as specified in RFC5424 if socket.gethostname() returns
    bad data (exception).

    This formatter will automatically insert the RFC5424 header for you;
    the format string that you pass in the constructor is only
    applied to the message body (and should typically just be %(message)).

    Stuctured Data Example:

        [exampleSDID@32473 iut="3" eventSource="Application" eventID="1011"]

    To use structured data:

       1. Construct the logger with an sd_id kwarg (or set the `sd_id` attribute on the logger object)
       2. Construct your individual records with `{'args': {'structured_data': {'iut': '3'}}}`
    '''
    def __init__(self, fmt='%(message)s', datefmt=None, style='%', procid=None, msgid=None, sd_id=None):
        # note: we accept and throw away "style" because our stuff only works with %
        # we also accept and throw away datefmt for similar reasons
        self._tz_fix = re.compile(r'([+-]\d{2})(\d{2})$')
        self._procid = procid
        self._msgid = msgid
        self._sd_id = sd_id
        return super(RFC5424Formatter, self).__init__(fmt=fmt, datefmt=None)

    @property
    def procid(self):
        """Default PROCID to add to syslog message"""
        return self._procid

    @procid.setter
    def procid(self, id):
        self._procid = id

    @property
    def msgid(self):
        """Default MSGID to add to syslog message"""
        return self._msgid

    @msgid.setter
    def msgid(self, id):
        self._msgid = id

    @property
    def sd_id(self):
        """Default SD-ID to add to STRUCTURED-DATA section in syslog message"""
        return self._sd_id

    @sd_id.setter
    def sd_id(self, sd_id):
        if not sd_id:
            raise InvalidSDIDError('SD-ID cannot be empty')
        self._sd_id = sd_id

    def format(self, record):
        try:
            record.__dict__['hostname'] = socket.gethostname()
        except Exception:
            record.__dict__['hostname'] = '-'
        isotime = datetime.datetime.fromtimestamp(record.created).isoformat()
        tz = self._tz_fix.match(time.strftime('%z'))
        if time.timezone and tz:
            (offset_hrs, offset_min) = tz.groups()
            if int(offset_hrs) == 0 and int(offset_min) == 0:
                isotime = isotime + 'Z'
            else:
                isotime = '{0}{1}:{2}'.format(isotime, offset_hrs, offset_min)
        else:
            isotime = isotime + 'Z'

        record.__dict__['isotime'] = isotime
        record.__dict__['procid'] = self.procid if self.procid else record.process
        record.__dict__['msgid'] = self.msgid if self.msgid else '-'

        if 'structured_data' in record.args:
            if not isinstance(record.args['structured_data'], dict):
                raise ValueError('structured_data must be a dict')

            all_sddata = {}
            default_sdparam = {}

            for key, value in record.args['structured_data'].items():
                if isinstance(value, dict):
                    all_sddata[key] = value
                else:
                    default_sdparam[key] = value

            if len(default_sdparam) > 0:
                if self.sd_id in all_sddata:
                    raise InvalidSDIDError('Cannot use same SD-ID twice')
                if not self.sd_id:
                    raise InvalidSDIDError('SD-ID cannot be empty')
                all_sddata[self.sd_id] = default_sdparam

            sd = ''
            for sdid, data in all_sddata.items():
                sd += '[{0}'.format(sdid)
                for key, value in data.items():
                    escaped = str(value).replace('\\', '\\\\').replace(']', '\\]').\
                        replace('"', '\\"')
                    sd += ' {0}="{1}"'.format(key, escaped)
                sd += ']'

            record.__dict__['sd'] = sd
        else:
            record.__dict__['sd'] = '-'

        for key in ('procid', 'msgid'):
            if key in record.args:
                record.__dict__[key] = record.args.pop(key)

        header = '1 {isotime} {hostname} {name} {procid} {msgid} {sd} '.format(
            **record.__dict__
        )
        return header + super(RFC5424Formatter, self).format(record)
